core/storage/trie/polkadot_trie_db/polkadot_trie_batch.cpp

```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "storage/trie/polkadot_trie_db/polkadot_trie_batch.hpp"
+
+namespace kagome::storage::trie {
+
+  using common::Buffer;
+
+  PolkadotTrieBatch::PolkadotTrieBatch(PolkadotTrieDb &trie) : trie_{trie} {}
+
+  outcome::result<void> PolkadotTrieBatch::put(const Buffer &key,
+                                               const Buffer &value) {
+    commands_.push_back({Action::PUT, key, value});
+    return outcome::success();
+  }
+
+  outcome::result<void> PolkadotTrieBatch::remove(const Buffer &key) {
+    commands_.push_back({Action::REMOVE, key});
+    return outcome::success();
+  }
+
+  outcome::result<void> PolkadotTrieBatch::commit() {
+    if (commands_.empty()) {
+      return outcome::success();
+    }
+
+    PolkadotTrieDb::NodePtr new_root = nullptr;
+    if (not trie_.getRootHash().empty()) {
+      OUTCOME_TRY(n, trie_.retrieveNode(trie_.getRootHash()));
+      new_root = n;
+    }
+    for (auto &command : commands_) {
+      switch (command.action) {
+        case Action::PUT: {
+          OUTCOME_TRY(
+              n, applyPut(new_root, command.key, std::move(command.value)));
```

Comment:
you are moving value from container which stays (at least till the end of function). If it exits before `clear()` commands_ becomes invalid. In this case I would take items from list one by one from beginning and remove them after execution of single command.
for example,
```
while(commands_.size() > 0) {
  auto & item = commands_.front();
  // ... do whatever you want
  commands_.pop_front();
}
```

Reply:
sorry my comment is not fully correct.
```
while(!commands_.is_empty()) {
  auto item = std::move(commands_.front());
  commands_.pop_front();
  // now do whatever you want :)
}
```

Reply:
You're completely right, and more than that I don't think that any commands at all should remain in a batch after a failure.